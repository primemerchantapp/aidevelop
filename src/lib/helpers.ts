import type { EmbedOptions, OpenOptions } from './interfaces';
import { DEFAULT_FRAME_HEIGHT, DEFAULT_ORIGIN } from './constants';
import { buildParams } from './params';

type Route = '/run' | `/edit/${string}` | `/github/${string}` | `/fork/${string}`;

/**
 * Pseudo-random id string for internal accounting.
 * 8 characters long, and collisions around 1 per million.
 */
export function genID() {
  return Math.random().toString(36).slice(2, 6) + Math.random().toString(36).slice(2, 6);
}

export function openUrl(route: Route, options?: OpenOptions) {
  return `${getOrigin(options)}${route}${buildParams(options)}`;
}

export function embedUrl(route: Route, options?: EmbedOptions) {
  const config: EmbedOptions = {
    forceEmbedLayout: true,
  };
  if (options && typeof options === 'object') {
    Object.assign(config, options);
  }
  return `${getOrigin(config)}${route}${buildParams(config)}`;
}

function getOrigin(options: OpenOptions & EmbedOptions = {}) {
  const origin = typeof options.origin === 'string' ? options.origin : DEFAULT_ORIGIN;
  return origin.replace(/\/$/, '');
}

export function replaceAndEmbed(
  target: HTMLElement,
  frame: HTMLIFrameElement,
  options?: EmbedOptions
) {
  if (!frame || !target || !target.parentNode) {
    throw new Error('Invalid Element');
  }
  if (target.id) {
    frame.id = target.id;
  }
  if (target.className) {
    frame.className = target.className;
  }
  setFrameDimensions(frame, options);
  target.replaceWith(frame);
}

export function findElement(elementOrId: string | HTMLElement) {
  if (typeof elementOrId === 'string') {
    const element = document.getElementById(elementOrId);
    if (!element) {
      throw new Error(`Could not find element with id '${elementOrId}'`);
    }
    return element;
  } else if (elementOrId instanceof HTMLElement) {
    return elementOrId;
  }
  throw new Error(`Invalid element: ${elementOrId}`);
}

export function openTarget(options?: OpenOptions) {
  return options && options.newWindow === false ? '_self' : '_blank';
}

function setFrameDimensions(frame: HTMLIFrameElement, options: EmbedOptions = {}) {
  const height: string = Object.hasOwnProperty.call(options, 'height')
    ? `${options.height}`
    : `${DEFAULT_FRAME_HEIGHT}`;
  const width: string | undefined = Object.hasOwnProperty.call(options, 'width')
    ? `${options.width}`
    : undefined;

  frame.setAttribute('height', height);
  if (width) {
    frame.setAttribute('width', width);
  } else {
    frame.setAttribute('style', 'width:100%;');
  }
}
