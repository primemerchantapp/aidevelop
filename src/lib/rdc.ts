import { genID } from './helpers';

interface MessagePayload {
  [key: string]: any;
}

interface MessagePayloadWithMetadata extends MessagePayload {
  __reqid: string;
  __success: boolean;
  __error?: string;
}

interface MessageData {
  type: string;
  payload: MessagePayloadWithMetadata;
}

interface RequestData {
  type: string;
  payload: MessagePayload;
}

interface PendingResolvers {
  [id: string]: {
    resolve(value: any): void;
    reject(error: string): void;
  };
}

export class RDC {
  private port: MessagePort;
  private pending: PendingResolvers = {};

  constructor(port: MessagePort) {
    this.port = port;
    this.port.onmessage = this.messageListener.bind(this);
  }

  public request<TResult = null>({ type, payload }: RequestData): Promise<TResult | null> {
    return new Promise((resolve, reject) => {
      const id = genID();
      this.pending[id] = { resolve, reject };
      this.port.postMessage({
        type,
        payload: {
          ...payload,
          // Ensure the payload object includes the request ID
          __reqid: id,
        },
      });
    });
  }

  private messageListener(event: MessageEvent<MessageData>) {
    if (typeof event.data.payload?.__reqid !== 'string') {
      return;
    }

    const { type, payload } = event.data;
    const { __reqid: id, __success: success, __error: error } = payload;

    if (this.pending[id]) {
      if (success) {
        this.pending[id].resolve(this.cleanResult(payload));
      } else {
        this.pending[id].reject(error ? `${type}: ${error}` : type);
      }
      delete this.pending[id];
    }
  }

  private cleanResult(payload: MessagePayloadWithMetadata): MessagePayload | null {
    const result: Partial<typeof payload> = { ...payload };
    delete result.__reqid;
    delete result.__success;
    delete result.__error;
    // Null the result if payload was empty besides the private metadata fields
    return Object.keys(result).length ? result : null;
  }
}
