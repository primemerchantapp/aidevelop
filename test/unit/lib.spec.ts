/** @vitest-environment happy-dom */
import { describe, expect, test, vi } from 'vitest';

import { connect, openGithubProject, openProject, openProjectId } from '$src/lib';
import { getTestProject } from './utils/project';

/**
 * Cannot test a working use case for connect or or all the embed* methods using happy-dom,
 * due to iframes not having a contentWindow and Window objects not having postMessage
 */
describe('connect', () => {
  test('rejects if target is not an iframe', async () => {
    const div = document.createElement('div');
    await expect(connect(div as any)).rejects.toBe(`Provided element is not an iframe.`);
  });
});

describe('openProject', () => {
  test('injects a form to open project in new page', () => {
    const project = getTestProject();

    // Spy on openProject doing DOM mutations
    let form: HTMLFormElement;
    const observerCb = vi.fn(([record]: MutationRecord[]) => {
      if (!record || form) return;
      record.addedNodes.forEach((node) => {
        if (node instanceof HTMLFormElement) {
          form = node;
        }
      });
    });
    const observer = new MutationObserver(observerCb);
    observer.observe(document.body, {
      childList: true,
      attributes: false,
      subtree: true,
    });

    expect(() => openProject(project)).not.toThrow();

    // We expect to have seen 2 DOM mutations (add then remove form)
    expect(observerCb).toHaveBeenCalledTimes(2);

    // We expect to have seen a form
    expect(form!).toBeInstanceOf(HTMLFormElement);
    expect(form!.attributes).toMatchSnapshot();

    observer.disconnect();
    observerCb.mockRestore();
  });
});

describe('openProjectId', () => {
  test('calls window.open', () => {
    const openSpy = vi.spyOn(window, 'open');

    openProjectId('js');
    expect(openSpy).toBeCalledWith('https://stackblitz.com/edit/js', '_blank');

    openProjectId('js', {
      origin: 'https://example.com',
      newWindow: false,
      openFile: 'src/index.js',
    });
    expect(openSpy).toBeCalledWith('https://example.com/edit/js?file=src%2Findex.js', '_self');

    openSpy.mockRestore();
  });
});

describe('openGithubProject', () => {
  test('calls window.open', () => {
    const openSpy = vi.spyOn(window, 'open');

    openGithubProject('stackblitz/sdk');
    expect(openSpy).toBeCalledWith('https://stackblitz.com/github/stackblitz/sdk', '_blank');

    openGithubProject('stackblitz/docs', { newWindow: false, hideExplorer: true });
    expect(openSpy).toBeCalledWith(
      'https://stackblitz.com/github/stackblitz/docs?hideExplorer=1',
      '_self'
    );

    openSpy.mockRestore();
  });
});
