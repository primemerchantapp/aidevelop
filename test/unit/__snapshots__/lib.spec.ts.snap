// Vitest Snapshot v1

exports[`openProject > injects a form to open project in new page 1`] = `
NamedNodeMap {
  "action": "https://stackblitz.com/run",
  "method": "POST",
  "style": "display:none!important;",
  "target": "_blank",
}
`;
